src/multipattern.rs

```diff
@@ -0,0 +1,270 @@
+    #[error(transparent)]
+    PatternParseError(E),
+    /// One of the clauses in the multipattern wasn't of the form `?var (= pattern)+`.
+    #[error("Bad clause in the multipattern: {0}")]
+    PatternAssignmentError(String),
+    /// One of the variables failed to parse.
+    #[error(transparent)]
+    VariableError(<Var as FromStr>::Err),
+}
+
+impl<L: Language + FromOp> FromStr for MultiPattern<L> {
+    type Err = MultiPatternParseError<<PatternAst<L> as FromStr>::Err>;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        use MultiPatternParseError::*;
+        let mut asts = vec![];
+        for split in s.trim().split(',') {
+            let split = split.trim();
+            if split.is_empty() {
+                continue;
+            }
+            let mut parts = split.split('=');
+            let vs: &str = parts
+                .next()
+                .ok_or_else(|| PatternAssignmentError(split.into()))?;
+            let v: Var = vs.trim().parse().map_err(VariableError)?;
+            let ps = parts
+                .map(|p| p.trim().parse())
+                .collect::<Result<Vec<PatternAst<L>>, _>>()
+                .map_err(PatternParseError)?;
+            if ps.is_empty() {
+                return Err(PatternAssignmentError(split.into()));
+            }
+            asts.extend(ps.into_iter().map(|p| (v, p)))
+        }
+        Ok(MultiPattern::new(asts))
+    }
+}
+
+impl<L: Language, A: Analysis<L>> Searcher<L, A> for MultiPattern<L> {
+    fn search_eclass(&self, egraph: &EGraph<L, A>, eclass: Id) -> Option<SearchMatches<L>> {
+        let substs = self.program.run(egraph, eclass);
+        if substs.is_empty() {
+            None
+        } else {
+            Some(SearchMatches {
+                eclass,
+                substs,
+                ast: None,
+            })
+        }
+    }
+
+    fn vars(&self) -> Vec<Var> {
+        let mut vars = vec![];
+        for (v, pat) in &self.asts {
+            vars.push(*v);
+            for n in pat.as_ref() {
+                if let ENodeOrVar::Var(v) = n {
+                    vars.push(*v)
+                }
+            }
+        }
+        vars.sort();
+        vars.dedup();
+        vars
+    }
+}
+
+impl<L: Language, A: Analysis<L>> Applier<L, A> for MultiPattern<L> {
+    fn apply_one(
+        &self,
+        _egraph: &mut EGraph<L, A>,
+        _eclass: Id,
+        _subst: &Subst,
+        _searcher_ast: Option<&PatternAst<L>>,
+        _rule_name: Symbol,
+    ) -> Vec<Id> {
+        panic!("Multipatterns do not support apply_one")
+    }
+
+    fn apply_matches(
+        &self,
+        egraph: &mut EGraph<L, A>,
+        matches: &[SearchMatches<L>],
+        _rule_name: Symbol,
+    ) -> Vec<Id> {
+        // TODO explanations?
+        // the ids returned are kinda garbage
+        let mut added = vec![];
+        for mat in matches {
+            for subst in &mat.substs {
+                let mut subst = subst.clone();
```

Comment:
This is an unfortunate clone, but I don't see an elegant way around it without changing interface of matches it be `&mut` or something.